File: server_elixir/test/afterlight/activities/hammer_strike_test.exs
defmodule Afterlight.Activities.HammerStrikeTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.HammerStrike

  describe "init_sim_state/1" do
    test "initializes a solo swinging hammer with five strikes" do
      state = HammerStrike.init_sim_state(slots: [0])

      assert state["status"] == "swinging"
      assert state["maxStrikes"] == 5
      assert state["remaining"] == 5
      assert state["strikes"] == []
      assert state["targetPhase"] == 0.5
      assert state["activeSlots"] == [0]
    end
  end

  describe "score_timing_sample/1" do
    test "identical samples produce identical score and bell" do
      assert {:ok, a} = HammerStrike.score_timing_sample(%{"phase" => 0.5})
      assert {:ok, b} = HammerStrike.score_timing_sample(%{"phase" => 0.5})
      assert a == b
      assert a["perfect"] == true
      assert a["score"] > 0.9
      assert a["bellHz"] > a["bellAmp"]
    end

    test "late samples ring lower and quieter than the target" do
      assert {:ok, perfect} = HammerStrike.score_timing_sample(%{"phase" => 0.5})
      assert {:ok, late} = HammerStrike.score_timing_sample(%{"phase" => 0.8})
      assert perfect["score"] > late["score"]
      assert perfect["bellHz"] > late["bellHz"]
      assert perfect["bellAmp"] > late["bellAmp"]
    end

    test "millisecond error is accepted and deterministic" do
      assert {:ok, a} = HammerStrike.score_timing_sample(%{"errorMs" => 40})
      assert {:ok, b} = HammerStrike.score_timing_sample(%{"errorMs" => 40})
      assert a == b
      assert {:ok, worse} = HammerStrike.score_timing_sample(%{"errorMs" => 400})
      assert a["score"] > worse["score"]
    end

    test "phase wins when both phase and errorMs are sent" do
      assert {:ok, mixed} = HammerStrike.score_timing_sample(%{"phase" => 0.5, "errorMs" => 800})
      assert {:ok, phase_only} = HammerStrike.score_timing_sample(%{"phase" => 0.5})
      assert mixed["score"] == phase_only["score"]
      assert mixed["bellHz"] == phase_only["bellHz"]
    end
  end

  describe "validate_controls/1" do
    test "rejects junk and unknown kinds" do
      assert {:error, :invalid_controls} = HammerStrike.validate_controls("nope")
      assert {:error, :unknown_kind} = HammerStrike.validate_controls(%{"kind" => "teleport"})
      assert {:ok, %{"kind" => "ready"}} = HammerStrike.validate_controls(%{"kind" => "ready"})
    end
  end

  describe "apply_input/3" do
    test "records a strike and ignores a duplicate commitId" do
      state = HammerStrike.init_sim_state()
      {state, event} = HammerStrike.apply_input(state, 0, %{"kind" => "strike", "phase" => 0.5, "commitId" => 1})

      assert event["type"] == "hammer_struck"
      assert length(state["strikes"]) == 1
      assert state["remaining"] == 4

      {again, dup} = HammerStrike.apply_input(state, 0, %{"kind" => "strike", "phase" => 0.1, "commitId" => 1})
      assert dup == nil
      assert length(again["strikes"]) == 1
    end

    test "five identical sequences complete with the same totals" do
      sequence = [0.5, 0.48, 0.62, 0.2, 0.51]

      {a, _} =
        Enum.reduce(Enum.with_index(sequence, 1), {HammerStrike.init_sim_state(), nil}, fn {phase, id}, {st, _} ->
          HammerStrike.apply_input(st, 0, %{"kind" => "strike", "phase" => phase, "commitId" => id})
        end)

      {b, _} =
        Enum.reduce(Enum.with_index(sequence, 1), {HammerStrike.init_sim_state(), nil}, fn {phase, id}, {st, _} ->
          HammerStrike.apply_input(st, 0, %{"kind" => "strike", "phase" => phase, "commitId" => id})
        end)

      assert a["status"] == "complete"
      assert b["status"] == "complete"
      assert Enum.map(a["strikes"], & &1["score"]) == Enum.map(b["strikes"], & &1["score"])
      assert a["totalScore"] == b["totalScore"]
      assert a["winner"] == 0
    end
  end

  describe "step_simulation/3" do
    test "advances phase and consumes a pending player strike once" do
      {state, _} = HammerStrike.step_simulation(HammerStrike.init_sim_state(), %{}, 30)
      assert state["phase"] > 0.0
      assert state["strikes"] == []

      players = %{0 => %{input_state: %{"kind" => "strike", "phase" => 0.5, "commitId" => 7}}}
      {state, event} = HammerStrike.step_simulation(state, players, 1)
      assert event == nil
      assert length(state["strikes"]) == 1

      {again, second} = HammerStrike.step_simulation(state, players, 1)
      assert second == nil
      assert length(again["strikes"]) == 1
    end

    test "completing the fifth strike emits match_ended for the session tick" do
      {state, _} =
        Enum.reduce(1..4, {HammerStrike.init_sim_state(), nil}, fn i, {st, _} ->
          HammerStrike.apply_input(st, 0, %{"kind" => "strike", "phase" => 0.5, "commitId" => i})
        end)

      players = %{0 => %{input_state: %{"kind" => "strike", "phase" => 0.51, "commitId" => 5}}}
      {state, event} = HammerStrike.step_simulation(state, players, 1)

      assert {:match_ended, 0, details} = event
      assert details["reason"] == "strikes_complete"
      assert state["status"] == "complete"
      assert state["resultEmitted"] == true
    end
  end
end

File: tests/hammer-strike.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  HAMMER_MAX_STRIKES,
  HAMMER_TARGET_PHASE,
  HAMMER_WINDOW_PHASE,
  wrapPhase,
  scoreTimingSample,
  validateHammerControls,
  initHammerSimState,
  applyHammerInput,
  stepHammerSimulation,
} from '../shared/hammerStrikeModel.js';
import { hasActivityModule } from '../src/activities/registry.js';
import '../src/activities/hammerStrike.js';

test('hammer-strike: module registers as hammer-strike', () => {
  assert.equal(hasActivityModule('hammer-strike'), true);
});

test('hammer-strike: wrapPhase folds samples onto [0, 1)', () => {
  assert.equal(wrapPhase(0.25), 0.25);
  assert.equal(wrapPhase(1.25), 0.25);
  assert.equal(wrapPhase(-0.25), 0.75);
  assert.equal(wrapPhase('nope'), null);
});

test('hammer-strike: identical timing samples score identically including bell', () => {
  const a = scoreTimingSample({ phase: 0.5 });
  const b = scoreTimingSample({ phase: 0.5 });
  assert.equal(a.valid, true);
  assert.deepEqual(a, b);
  assert.equal(a.perfect, true);
  assert.ok(a.score > 0.9);
  assert.ok(a.bellHz > a.bellAmp);
});

test('hammer-strike: early and late samples are worse than the target window', () => {
  const perfect = scoreTimingSample({ phase: HAMMER_TARGET_PHASE });
  const early = scoreTimingSample({ phase: HAMMER_TARGET_PHASE - HAMMER_WINDOW_PHASE * 3 });
  const late = scoreTimingSample({ phase: HAMMER_TARGET_PHASE + 0.3 });
  assert.ok(perfect.score > early.score);
  assert.ok(perfect.score > late.score);
  assert.ok(perfect.bellHz > late.bellHz);
  assert.ok(perfect.bellAmp > late.bellAmp);
});

test('hammer-strike: millisecond error is accepted and deterministic', () => {
  const a = scoreTimingSample({ errorMs: 40 });
  const b = scoreTimingSample({ errorMs: 40 });
  assert.equal(a.valid, true);
  assert.deepEqual(a, b);
  const worse = scoreTimingSample({ errorMs: 400 });
  assert.ok(a.score > worse.score);
});

test('hammer-strike: phase wins when both phase and errorMs are sent', () => {
  const mixed = scoreTimingSample({ phase: 0.5, errorMs: 800 });
  const phaseOnly = scoreTimingSample({ phase: 0.5 });
  assert.deepEqual(mixed, phaseOnly);
});

test('hammer-strike: validateHammerControls rejects junk and unknown kinds', () => {
  assert.equal(validateHammerControls(null).valid, false);
  assert.equal(validateHammerControls({ kind: 'teleport' }).valid, false);
  const ready = validateHammerControls({ kind: 'ready' });
  assert.equal(ready.valid, true);
  assert.equal(ready.sanitized.kind, 'ready');
});

test('hammer-strike: applyHammerInput records a strike and ignores duplicate commitId', () => {
  let state = initHammerSimState();
  assert.equal(state.remaining, HAMMER_MAX_STRIKES);

  const first = applyHammerInput(state, 0, { kind: 'strike', phase: 0.5, commitId: 1 });
  state = first.simState;
  assert.equal(first.event.type, 'hammer_struck');
  assert.equal(state.strikes.length, 1);
  assert.equal(state.remaining, HAMMER_MAX_STRIKES - 1);

  const dup = applyHammerInput(state, 0, { kind: 'strike', phase: 0.1, commitId: 1 });
  assert.equal(dup.event, null);
  assert.equal(dup.simState.strikes.length, 1);
});

test('hammer-strike: five strikes complete with deterministic totals', () => {
  const sequence = [0.5, 0.48, 0.62, 0.2, 0.51];
  let a = initHammerSimState();
  let b = initHammerSimState();

  for (let i = 0; i < sequence.length; i++) {
    a = applyHammerInput(a, 0, { kind: 'strike', phase: sequence[i], commitId: i + 1 }).simState;
    b = applyHammerInput(b, 0, { kind: 'strike', phase: sequence[i], commitId: i + 1 }).simState;
  }

  assert.equal(a.status, 'complete');
  assert.equal(b.status, 'complete');
  assert.deepEqual(a.strikes.map((s) => s.score), b.strikes.map((s) => s.score));
  assert.equal(a.totalScore, b.totalScore);
  assert.equal(a.bestScore, b.bestScore);
  assert.equal(a.winner, 0);
});

test('hammer-strike: step advances phase and consumes a pending player strike once', () => {
  let { simState } = stepHammerSimulation(initHammerSimState(), {}, 30);
  assert.ok(simState.phase > 0);
  assert.equal(simState.strikes.length, 0);

  const players = { 0: { input_state: { kind: 'strike', phase: 0.5, commitId: 7 } } };
  const first = stepHammerSimulation(simState, players, 1);
  assert.equal(first.event?.type, 'hammer_struck');
  assert.equal(first.simState.strikes.length, 1);

  const again = stepHammerSimulation(first.simState, players, 1);
  assert.notEqual(again.event?.type, 'hammer_struck');
  assert.equal(again.simState.strikes.length, 1);
});
